domain/charm/state/state.go

```diff
@@ -1097,12 +1120,86 @@ ORDER BY array_index ASC;
 	return result, nil
 }
 
+func (s *State) checkSetCharmExists(ctx context.Context, tx *sqlair.TX, id corecharm.ID, name string, revision int) error {
+	selectQuery := `
+SELECT charm.uuid AS &charmID.*
+FROM charm
+LEFT JOIN charm_origin ON charm.uuid = charm_origin.charm_uuid
+WHERE charm.name = $charmNameRevision.name AND charm_origin.revision = $charmNameRevision.revision
+
+	`
+	selectStmt, err := s.Prepare(selectQuery, charmID{}, charmNameRevision{})
+	if err != nil {
+		return fmt.Errorf("failed to prepare query: %w", err)
+	}
+	if err := tx.Query(ctx, selectStmt, charmNameRevision{
+		Name:     name,
+		Revision: revision,
+	}).Get(&charmID{}); err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return fmt.Errorf("failed to check charm exists: %w", err)
+	}
+
+	return charmerrors.AlreadyExists
+}
+
+func (s *State) setCharmHash(ctx context.Context, tx *sqlair.TX, id corecharm.ID, hash string) error {
+	ident := charmID{UUID: id.String()}
+
+	query := `INSERT INTO charm_hash (*) VALUES ($setCharmHash.*);`
+	stmt, err := s.Prepare(query, setCharmHash{})
+	if err != nil {
+		return fmt.Errorf("failed to prepare query: %w", err)
+	}
+
+	if err := tx.Query(ctx, stmt, setCharmHash{
+		CharmUUID:  ident.UUID,
+		HashKindID: sha256HashKind,
+		Hash:       hash,
+	}).Run(); err != nil {
+		return fmt.Errorf("failed to insert charm hash: %w", err)
+	}
+
+	return nil
+}
+
+func (s *State) setCharmInitialOrigin(
+	ctx context.Context, tx *sqlair.TX, id corecharm.ID,
+	source charm.CharmSource, revision int, version string) error {
+	ident := charmID{UUID: id.String()}
+
+	encodedOriginSource, err := encodeOriginSource(source)
+	if err != nil {
+		return fmt.Errorf("failed to encode charm origin source: %w", err)
+	}
+
+	query := `INSERT INTO charm_origin (*) VALUES ($setCharmSourceRevisionVersion.*);`
+	stmt, err := s.Prepare(query, setCharmSourceRevisionVersion{})
```

Comment:
Can declar the `setCharmSourceRevisionVersion` struct before the prepare and reuse.

Also could the name be made more concise?

Reply:
It's the charm source revision and version, as these are row-based column types, I want the names of the types to represent that.